runners/direct-java/src/main/java/org/apache/beam/runners/direct/AggregatorContainer.java

```diff
@@ -0,0 +1,154 @@
+import javax.annotation.Nullable;
+
+/**
+ * AccumT container for the current values associated with {@link Aggregator Aggregators}.
+ */
+public class AggregatorContainer {
+
+  private static class AggregatorInfo<InputT, AccumT, OutputT>
+      implements Aggregator<InputT, OutputT> {
+    private final String name;
+    private final CombineFn<InputT, AccumT, OutputT> combiner;
+    private AccumT accumulator = null;
+    private boolean committed = false;
+
+    private AggregatorInfo(String name, CombineFn<InputT, AccumT, OutputT> combiner) {
+      this.name = name;
+      this.combiner = combiner;
+    }
+
+    @Override
+    public void addValue(InputT input) {
+      Preconditions.checkState(!committed, "Cannot addValue after committing");
+      if (accumulator == null) {
+        accumulator = combiner.createAccumulator();
+      }
+      accumulator = combiner.addInput(accumulator, input);
+    }
+
+    public OutputT getOutput() {
+      return accumulator == null ? null : combiner.extractOutput(accumulator);
+    }
+
+    private void merge(AggregatorInfo<?, ?, ?> other) {
+      // TODO: Make sure the CombineFn and accumulators are compatible.
+      AggregatorInfo<InputT, AccumT, OutputT> otherSafe =
+          (AggregatorInfo<InputT, AccumT, OutputT>) other;
+      mergeSafe(otherSafe);
+    }
+
+    private void mergeSafe(AggregatorInfo<InputT, AccumT, OutputT> other) {
+      if (accumulator == null) {
+        accumulator = other.accumulator;
+      } else if (other.accumulator != null) {
+        accumulator = combiner.mergeAccumulators(Arrays.asList(accumulator, other.accumulator));
+      }
+    }
+
+    @Override
+    public String getName() {
+      return name;
+    }
+
+    @Override
+    public CombineFn<InputT, ?, OutputT> getCombineFn() {
+      return combiner;
+    }
+  }
+
+  private final HashMap<String, AggregatorInfo<?, ?, ?>> accumulators = new HashMap<>();
+
+  private AggregatorContainer() {
+  }
+
+  public static AggregatorContainer create() {
+    return new AggregatorContainer();
+  }
+
+  @Nullable
+  <OutputT> OutputT getAggregate(String name) {
+    AggregatorInfo<?, ?, OutputT> aggregatorInfo =
+        (AggregatorInfo<?, ?, OutputT>) accumulators.get(name);
+    return aggregatorInfo == null ? null : aggregatorInfo.getOutput();
+  }
+
+  public Mutator createMutator() {
+    return new Mutator(this);
+  }
+
+  /**
+   * AccumT class for mutations to the aggregator values.
+   */
+  public static class Mutator implements AggregatorFactory {
+
+    private final HashMap<String, AggregatorInfo<?, ?, ?>> accumulatorDeltas = new HashMap<>();
+    private final AggregatorContainer container;
+    private boolean committed = false;
+
+    private Mutator(AggregatorContainer container) {
+      this.container = container;
+    }
+
+    public void commit() {
+      Preconditions.checkState(!committed, "Should not be already committed");
+      committed = true;
+      synchronized (container) {
```

Comment:
If you do decide to use `ConcurrentMap`, you should instead synchronize on the existing `AggregatorInfo`